cmd/reconciler/main.go

```diff
@@ -161,68 +399,97 @@ func main() {
-		StatusUpdatePeriod:      configsync.DefaultReconcilerSyncStatusUpdatePeriod,
-		SourceRoot:              absSourceDir,
-		RepoRoot:                absRepoRoot,
-		HydratedRoot:            *hydratedRootDir,
-		HydratedLink:            *hydratedLinkDir,
-		SourceRev:               *sourceRev,
-		SourceBranch:            *sourceBranch,
-		SourceType:              v1beta1.SourceType(*sourceType),
-		SourceRepo:              *sourceRepo,
-		SyncDir:                 relSyncDir,
-		SyncName:                *syncName,
-		ReconcilerName:          *reconcilerName,
-		StatusMode:              *statusMode,
-		ReconcileTimeout:        *reconcileTimeout,
-		APIServerTimeout:        *apiServerTimeout,
-		RenderingEnabled:        *renderingEnabled,
-	}
-
-	if declared.Scope(*scope) == declared.RootReconciler {
+		return nil, fmt.Errorf("%s must be an absolute path: %v", flags.sourceDir, err)
+	}
+
+	if err = declared.ValidateScope(*scope); err != nil {
+		return nil, err
+	}
+
+	var format filesystem.SourceFormat
+	var nsStrat configsync.NamespaceStrategy
+	switch declared.Scope(*scope) {
+	case declared.RootReconciler:
+		klog.Info("Starting reconciler for: root")
 		// Default to "hierarchy" if unset.
-		format := filesystem.SourceFormat(*sourceFormat)
+		format = filesystem.SourceFormat(*sourceFormat)
 		if format == "" {
 			format = filesystem.SourceFormatHierarchy
 		}
 		// Default to "implicit" if unset.
-		nsStrat := configsync.NamespaceStrategy(*namespaceStrategy)
+		nsStrat = configsync.NamespaceStrategy(*namespaceStrategy)
 		if nsStrat == "" {
 			nsStrat = configsync.NamespaceStrategyImplicit
 		}
-
-		klog.Info("Starting reconciler for: root")
-		opts.RootOptions = &reconciler.RootOptions{
-			SourceFormat:      format,
-			NamespaceStrategy: nsStrat,
-		}
-	} else {
+	default:
 		klog.Infof("Starting reconciler for: %s", *scope)
-
 		if *sourceFormat != "" {
-			klog.Fatalf("Flag %s and environment variable %s must not be passed to a Namespace reconciler",
+			return nil, fmt.Errorf("flag %s and environment variable %s must not be passed to a Namespace reconciler",
 				flags.sourceFormat, filesystem.SourceFormatKey)
 		}
 		if *namespaceStrategy != "" {
-			klog.Fatalf("Flag %s and environment variable %s must not be passed to a Namespace reconciler",
+			return nil, fmt.Errorf("flag %s and environment variable %s must not be passed to a Namespace reconciler",
 				flags.namespaceStrategy, reconcilermanager.NamespaceStrategy)
 		}
 	}
-	reconciler.Run(opts)
+
+	var parser parse.Parser
+	fs := parse.FileSource{
+		SourceDir:    absSourceDir,
+		RepoRoot:     absRepoRoot,
+		HydratedRoot: *hydratedRootDir,
+		HydratedLink: *hydratedLinkDir,
+		SyncDir:      relSyncDir,
+		SourceType:   v1beta1.SourceType(*sourceType),
+		SourceRepo:   *sourceRepo,
+		SourceBranch: *sourceBranch,
+		SourceRev:    *sourceRev,
+	}
+	ds := declared.Scope(*scope)
+	if ds == declared.RootReconciler {
+		parser, err = parse.NewRootRunner(
```

Comment:
these constructors are crazy long. perhaps we should replace them with a builder so we can just populate one builder and use it to build both runners.